core/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java

```diff
@@ -126,71 +119,20 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-                        }
-                    }
-                } else if (parseContext.getParseFieldMatcher().match(currentFieldName, TYPE_FIELD)) {
-                    while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
-                        String value = parser.textOrNull();
-                        if (value == null) {
-                            throw new ParsingException(parser.getTokenLocation(), "No type specified for term filter");
-                        }
-                        types.add(value);
-                    }
-                } else {
-                    throw new ParsingException(parser.getTokenLocation(), "[" + IdsQueryBuilder.NAME +
-                            "] query does not support [" + currentFieldName + "]");
-                }
-            } else if (token.isValue()) {
-                if (parseContext.getParseFieldMatcher().match(currentFieldName, TYPE_FIELD)) {
-                    types = Collections.singletonList(parser.text());
-                } else if (parseContext.getParseFieldMatcher().match(currentFieldName, AbstractQueryBuilder.BOOST_FIELD)) {
-                    boost = parser.floatValue();
-                } else if (parseContext.getParseFieldMatcher().match(currentFieldName, AbstractQueryBuilder.NAME_FIELD)) {
-                    queryName = parser.text();
-                } else {
-                    throw new ParsingException(parser.getTokenLocation(), "[" + IdsQueryBuilder.NAME +
-                            "] query does not support [" + currentFieldName + "]");
-                }
-            } else {
-                throw new ParsingException(parser.getTokenLocation(), "[" + IdsQueryBuilder.NAME +
-                        "] unknown token [" + token + "] after [" + currentFieldName + "]");
-            }
-        }
-        if (!idsProvided) {
-            throw new ParsingException(parser.getTokenLocation(), "[" + IdsQueryBuilder.NAME + "] query, no ids values provided");
-        }
+    @SuppressWarnings("unchecked")
+    private static ConstructingObjectParser<IdsQueryBuilder, QueryParseContext> PARSER = new ConstructingObjectParser<>(NAME,
+            a -> new IdsQueryBuilder(((List<String>) a[0]).toArray(new String[((List<String>) a[0]).size()])));
 
-        IdsQueryBuilder query = new IdsQueryBuilder(types.toArray(new String[types.size()]));
-        query.addIds(ids.toArray(new String[ids.size()]));
-        query.boost(boost).queryName(queryName);
-        return Optional.of(query);
+    static {
+        PARSER.declareStringArray(constructorArg(), IdsQueryBuilder.TYPE_FIELD);
+        PARSER.declareStringArray((builder, values) -> builder.addIds(values.toArray(new String[values.size()])),
+                IdsQueryBuilder.VALUES_FIELD);
+        PARSER.declareFloat((builder, value) -> builder.boost(value), AbstractQueryBuilder.BOOST_FIELD);
+        PARSER.declareString((builder, value) -> builder.queryName(value), AbstractQueryBuilder.NAME_FIELD);
```

Comment:
We might should move these last two declarations to a common spot something like
```
static <T extends AbstractObjectParser<? extends QueryBuilder>> declareStandardFields(T parser) {
        parser.declareFloat((builder, value) -> builder.boost(value), AbstractQueryBuilder.BOOST_FIELD);
        parser.declareString((builder, value) -> builder.queryName(value), AbstractQueryBuilder.NAME_FIELD);
        return parser;
}
```

and then we can declare them when we're initializing the object.